### PR TITLE
[Merged by Bors] - feat(AlgebraicGeometry/Tilde): add the map from `M` localising at `x` to the stalk of `M^~` at x

### PR DESCRIPTION
**Pull request by jjaassoonn**
Co-authored-by: Weihong Xu <weihong@vt.edu>
Co-authored-by: Sophie Morel <sophie.morel@ens-lyon.fr>
Co-authored-by: Amelia Livingston <al3717@ic.ac.uk>

This contribution was created as part of the AIM workshop "Formalizing algebraic geometry" in
June 2024.

---

- [x] depends on: #15315

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Limits/ConcreteCategory.lean [hunk 1]
```diff
@@ -144,6 +146,42 @@ theorem Concrete.colimit_rep_eq_iff_exists [HasColimit F] {i j : J} (x : F.obj i
...
 #align category_theory.limits.concrete.colimit_rep_eq_iff_exists CategoryTheory.Limits.Concrete.colimit_rep_eq_iff_exists
 
+theorem Concrete.colimit_rep_eq_zero
+    [∀ c : C, Zero c] [∀ {c c' : C}, ZeroHomClass (c ⟶ c') c c'] [HasColimit F]
+    (j : J) (x : F.obj j) (hx : colimit.ι F j x = 0) :
+    ∃ (j' : J) (i : j ⟶ j'), F.map i x = 0 := by
+  rw [show (0 : (forget C).obj (colimit F)) = colimit.ι F j 0 by simp,
+    Concrete.colimit_rep_eq_iff_exists] at hx
+  obtain ⟨j', i, y, g⟩ := hx
+  exact ⟨j', i, g ▸ by simp⟩
```
Comment (jjaassoonn) [inline Mathlib/CategoryTheory/Limits/ConcreteCategory.lean:156]: This lemma is no longer useful in this PR, but maybe we can keep it?

File: Mathlib/CategoryTheory/Limits/ConcreteCategory.lean [hunk 2]
```diff
@@ -144,6 +146,37 @@ theorem Concrete.colimit_rep_eq_iff_exists [HasColimit F] {i j : J} (x : F.obj i
...
+
+/--
+if `r` has no zero smul divisors for all small-enough sections, then `r` has no zero smul divisors
+on the level of stalks.
```
Comment (kbuzzard) [inline Mathlib/CategoryTheory/Limits/ConcreteCategory.lean:160]:
```suggestion
in the colimit.
```

File: Mathlib/CategoryTheory/Limits/ConcreteCategory.lean [hunk 3]
```diff
@@ -144,6 +146,37 @@ theorem Concrete.colimit_rep_eq_iff_exists [HasColimit F] {i j : J} (x : F.obj i
...
+  obtain ⟨j', i, h⟩ := Concrete.colimit_rep_eq_zero (hx := hx)
+  obtain ⟨j'', H⟩ := H
+  let s : J := IsFiltered.sup {j, j', j''} { ⟨j, j', by simp, by simp, i⟩ }
+  replace H := H s (IsFiltered.toSup _ _ $ by simp) (F.map (IsFiltered.toSup _ _ $ by simp) x)
+  rw [← LinearMapClass.map_smul, ← IsFiltered.toSup_commutes, F.map_comp, comp_apply, h, map_zero,
+    ← F.map_comp, IsFiltered.toSup_commutes] at H
+  have := congr(colimit.ι F _ $(H rfl))
+  all_goals try simp
+  simp only [elementwise_of% (colimit.w F), map_zero] at this
+  aesop
```
Comment (kbuzzard) [inline Mathlib/CategoryTheory/Limits/ConcreteCategory.lean:178]:
```suggestion
  aesop -- **TODO** this is a workaround for a tactic bug; `exact this` should work.
```

File: Mathlib/AlgebraicGeometry/Modules/Tilde.lean [hunk 4]
```diff
@@ -171,4 +178,94 @@ noncomputable def tilde : (Spec (CommRingCat.of R)).Modules where
     map_smul := fun _ _ _ => rfl }
   isSheaf := M.tildeInAddCommGrp.2
 
+/--
+This is `M^~` as a sheaf of `R`-modules.
+-/
+noncomputable def tildeInModuleCat :
+    TopCat.Presheaf (ModuleCat R) (PrimeSpectrum.Top R) :=
+  (PresheafOfModules.forgetToPresheafModuleCat (op ⊤) $
```
Comment (kbuzzard) [inline Mathlib/AlgebraicGeometry/Modules/Tilde.lean:186]: I thought the fashion was `<|` now, not `$`? But I might be wrong.

File: Mathlib/CategoryTheory/Limits/ConcreteCategory.lean [hunk 5]
```diff
@@ -7,6 +7,8 @@ import Mathlib.CategoryTheory.ConcreteCategory.Basic
 import Mathlib.CategoryTheory.Limits.Preserves.Basic
 import Mathlib.CategoryTheory.Limits.TypesFiltered
 import Mathlib.CategoryTheory.Limits.Yoneda
+import Mathlib.Algebra.Module.LinearMap.Defs
```
Comment (kbuzzard) [inline Mathlib/CategoryTheory/Limits/ConcreteCategory.lean:10]: A consequence of this import is that now the file proving that the category of abelian groups is abelian, now imports a bunch of ring theory. Maybe move this theorem to a new file?

**Review by kbuzzard — APPROVED** [review]
Modulo the added import, which I think should be fixed by moving a lemma to a new file, this looks great to me! Looking forward to seeing the proof that the stalk is M_x.

**Comment by erdOne** [comment]
I personally think that the added import is fine as the file is a leaf file and the contents added are core properties of tilde M. But any maintainer might very well disagree.
Otherwise LGTM
maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**